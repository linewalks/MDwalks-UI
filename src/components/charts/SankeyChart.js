import React from 'react'
import * as d3Core from 'd3'
import PropTypes from 'prop-types'
import * as sankeyCircular from 'd3-sankey-circular'
import _ from 'lodash'
import { color } from '@src/assets/styles/variables'

import { strIdConvert } from '@src/helper/chartUtility'

class SankeyChart extends React.Component {
  rootElement = React.createRef()

  constructor(props) {
    super(props)
    this.d3 = { ...d3Core, ...sankeyCircular }
    this.mapLinks = new Map()
    this.state = {
      selectedNodes: props.selectedNodes || [],
    }
  }

  getRootElement() {
    return this.d3.select(this.rootElement.current)
  }

  getNodeName = (node) => node.name

  setSelectedNode = (_selectedNode) => {
    const { selectedNodes } = this.state
    this.setState({
      selectedNodes: _.uniq(selectedNodes.concat(_selectedNode)),
    })
  }

  highlightLink = () => {
    const { selectedNodes } = this.state
    this.d3.selectAll(`.sankey-link`).style('opacity', 0.04).style('stroke', '#000000')

    const ids = this.createLinkId(selectedNodes)

    for (let i = 0; i < ids.length; i += 1) {
      const [source, target] = ids[i].split('X');
      const forwardPath = this.getRootElement().select(`#${source}X${target}`);
      const reversePath = this.getRootElement().select(`#${target}X${source}`);

      const sourceXPosition = this.getRootElement().select(`#${source}`).attr('x')
      const targetXPosition = this.getRootElement().select(`#${target}`).attr('x')

      if (targetXPosition > sourceXPosition) {
        forwardPath.style('opacity', 1).style('stroke', color.$pathway_link_blue)
      } else {
        reversePath.style('opacity', 1).style('stroke', color.$pathway_link_red)
      }
    }
  }

  createLinkId = (selectedNodes) => selectedNodes.map((node, i) => `${strIdConvert([node, selectedNodes[i + 1] || ''])}`).slice(0, -1)

  renderPlaceholder = () => <div>No data is provided!</div>

  initializeSankey = ({
    nodeWidth, width, height, iterations, circularLinkGap,
  }) => this.d3
    .sankeyCircular()
    .nodeWidth(nodeWidth)
    .nodePaddingRatio(0.5)
    .size([width, height])
    .nodeId((d) => d.name)
    .iterations(iterations)
    .circularLinkGap(circularLinkGap)

  initializeSVG = ({ width, height, margin }) => this.d3
    .select(this.rootElement.current)
    .append('svg')
    .attr('width', width + margin.left + margin.right)
    .attr('height', height + margin.top + margin.bottom)

  initializeGroups = (svg, { margin }) => {
    const entireGroup = svg
      .append('g')
      .attr('transform', `translate(${margin.left}, ${margin.top})`)

    const nodeGroup = entireGroup
      .append('g')
      .attr('class', 'nodes')
      .attr('font-family', 'sans-serif')
      .attr('font-size', 10)
      .selectAll('g')

    const linkGroup = entireGroup
      .append('g')
      .attr('class', 'links')
      .attr('fill', 'none')
      .selectAll('path')

    return [nodeGroup, linkGroup]
  }

  renderNodes = (nodeG, sankeyNodesData, { width }) => {
    this.d3
      .scaleSequential(this.d3.interpolateCool)
      .domain([0, width])
    // Enter node g
    const node = nodeG
      .data(sankeyNodesData)
      .enter()
      .append('g')
    // Render rect
    node
      .append('rect')
      .attr('x', (d) => d.x0)
      .attr('y', (d) => d.y0)
      .attr('height', (d) => d.y1 - d.y0)
      .attr('width', (d) => d.x1 - d.x0)
      .attr('rx', 4)
      .attr('ry', 4)
      .attr('id', (d) => strIdConvert(d.name))
      .style('fill', '#002d4f')
      .style('cursor', 'pointer')

    // Render node name
    node
      .append('text')
      .attr('x', (d) => (d.x0 + d.x1) / 2)
      .attr('y', (d) => d.y0 - 12)
      .attr('dy', '0.35em')
      .attr('fill', 'rgba(0, 0, 0, 0.4)')
      .attr('font-famliy', 'Spoqa Han Sans', 'Spoqa Han Sans JP', 'Sans-serif')
      .attr('text-anchor', 'middle')
      .attr('font-size', '14')
      .attr('font-weight', 'normal')
      .attr('font-style', 'normal')
      .attr('font-stretch', 'normal')
      .attr('line-height', 'normal')
      .attr('letter-spacing', '0.5px')
      .text((d) => d.name)
    /*
    pr url:https://github.com/linewalks/Cardio_Demo_View/pull/52/files
    TODO: node tooltip 사용성이 확정되면 다시 기능 추가할것.
    Render node title
    node.append('title').text(d => `${d.name}\n${d.value}`)
    */
    return node
  }

  renderLinks = (linkG, sankeyLinksData) => {
    const link = linkG
      .data(sankeyLinksData)
      .enter()
      .append('g')

    const { mapLinks } = this

    link
      .append('path')
      .attr('class', 'sankey-link')
      .attr('id', function ({ source, target }) {
        const id = `${strIdConvert([source.name, target.name])}`
        mapLinks.set(id, this)
        return id
      })
      .attr('d', ({ path }) => path)
      .style('stroke-width', (d) => Math.max(1, d.width))
      .style('opacity', 0.04)
      .style('stroke', '#000000') // reset 과 동일

    this.mapLinks = mapLinks
    // link.selectAll(`.sankey-link`).style('opacity', 0.04).style('stroke', '#000000')
    /*
    pr url:https://github.com/linewalks/Cardio_Demo_View/pull/52/files
    TODO: link tooltip 사용성이 확정되면 다시 기능 추가할것.
    link.append('title').text(d => {
      return `${d.source.name} → ${d.target.name}\n Count: ${d.value}`
    })
    */
    return link
  }

  linkConnectCheck = (prevSelectedNode, currentSelectedNode) => {
    if (_.isEmpty(prevSelectedNode)) {
      return true
    }

    const key1 = strIdConvert([prevSelectedNode, currentSelectedNode])
    const key2 = strIdConvert([currentSelectedNode, prevSelectedNode])

    return this.mapLinks.has(key1) || this.mapLinks.has(key2)
  }

  renderSankey = () => {
    const { d3 } = this
    const { options, data, onChange } = this.props

    const {
      height,
      width,
      margin,
      nodeWidth,
      // nodePadding,
      iterations,
      circularLinkGap,
    } = options

    this.sankey = this.initializeSankey({
      height,
      width,
      nodeWidth,
      iterations,
      circularLinkGap,
    })

    this.svg = this.initializeSVG({
      width,
      height,
      margin,
    })

    // initialize entire group, link group, node group
    const [nodeG, linkG] = this.initializeGroups(this.svg, { margin })

    const sankeyData = this.sankey(data)

    const nodes = this.renderNodes(nodeG, sankeyData.nodes, { width })
    this.renderLinks(linkG, sankeyData.links)

    this.attachEventHandlersToNode(d3, nodes, {
      onChange,
    })
  }

  attachEventHandlersToNode = (d3, nodes) => {
    nodes.on('click', (node) => {
      const { selectedNodes } = this.state
      const prevSelectedNode = _.last(selectedNodes)
      const currentSelectedNode = this.getNodeName(node)
      if (this.linkConnectCheck(prevSelectedNode, currentSelectedNode)) {
        this.setSelectedNode(this.getNodeName(node))
      }
    })
  }

  removeSankey = () => (
    this.getRootElement().select('svg').remove()
  )

  componentDidMount = () => {
    const { data, resetBtnId } = this.props
    if (!_.isEmpty(resetBtnId)) {
      this.d3.select(`#${resetBtnId}`).on('click', this.resetSankey)
    }

    if (!_.isEmpty(data)) {
      this.renderSankey()
      this.highlightLink()
    }
  }

  componentDidUpdate = (prevProps, prevState) => {
    const { data, onChange } = this.props
    const { selectedNodes } = this.state

    if (!_.isEqual(prevProps.data, data)) {
      this.removeSankey()
      this.resetSankey()
      return this.renderSankey()
    }

    if (JSON.stringify(selectedNodes) !== JSON.stringify(prevState.selectedNodes)) {
      onChange(selectedNodes)
    }
    this.highlightLink()
    return null
  }

  resetSankey = () => {
    const { defaultdNode } = this.props
    this.setState({ selectedNodes: defaultdNode })
  }

  render() {
    const { data } = this.props
    return _.isEmpty(data) ? (
      this.renderPlaceholder()
    ) : (
      <div ref={this.rootElement} />
    )
  }
}

SankeyChart.defaultProps = {
  selectedNodes: [],
  defaultdNode: [],
  onChange: () => {},
  options: {
    height: 254,
    width: 1000,
    nodeWidth: 25,
    nodePadding: 20,
    iterations: 15,
    circularLinkGap: 1,
    margin: {
      top: 100,
      right: 100,
      bottom: 100,
      left: 100,
    },
  },
  resetBtnId: undefined,
}

SankeyChart.propTypes = {
  selectedNodes: PropTypes.arrayOf(PropTypes.any),
  defaultdNode: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func,
  data: PropTypes.shape({
    links: PropTypes.arrayOf(PropTypes.shape()),
    nodes: PropTypes.arrayOf(PropTypes.shape()),
  }).isRequired,
  options: PropTypes.shape({
    height: PropTypes.number,
    width: PropTypes.number,
    margin: PropTypes.shape({
      top: PropTypes.number,
      right: PropTypes.number,
      bottom: PropTypes.number,
      left: PropTypes.number,
    }),
    nodeWidth: PropTypes.number,
    iterations: PropTypes.number,
    circularLinkGap: PropTypes.number,
  }),
  resetBtnId: PropTypes.string,
}

export default SankeyChart
