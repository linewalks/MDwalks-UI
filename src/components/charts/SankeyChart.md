SankeyChart example:

```js

import Button from '@Components/button/Button';

const SankeyChartExample = () => (
  <>
    <Button id="resetSankeyChart">reset</Button>
    <SankeyChart
      resetBtnId={"resetSankeyChart"}
      onChange={(nodes) => {console.log(nodes)}}
      selectedNodes={['ER Admission', 'Re-operation']}
      data={
        {
          "links": [
            {
              "source": "ER Admission",
              "target": "Re-operation",
              "value": 683
            },
            {
              "source": "Index Invasive Treatment",
              "target": "ER Death",
              "value": 44
            },
            {
              "source": "Re-operation",
              "target": "In Hospital Death",
              "value": 84
            },
            {
              "source": "Re-operation",
              "target": "ER Death",
              "value": 9
            },
          ],
          "nodes": [
            {
              "name": "Index Invasive Treatment"
            },
            {
              "name": "In Hospital Death"
            },
            {
              "name": "ER Admission"
            },
            {
              "name": "Re-operation"
            },
            {
              "name": "ER Death"
            },
          ]
        }
      }
    />
  </>
)

SankeyChartExample()
 ```