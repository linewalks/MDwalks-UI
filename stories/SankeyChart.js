import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs } from '@storybook/addon-knobs';
import SankeyChart from '@Charts/SankeyChart';
import sankeyData from '@Data/dataForSankey';

const eventHandler = (selectnodes) => {
  console.log(selectnodes)
}

storiesOf('Sankey Chart Component', module)
  .addDecorator(withKnobs)
  .add('basic', () => (
    <SankeyChart
      data={sankeyData}
    />
  ))
  .add('handles onNodeClick events', () => (
    <SankeyChart
      data={sankeyData}
      onNodeClick = {
        eventHandler
      }
    />
  ))
  .add('handles onLinkClick events', () => (
    <SankeyChart
      data={sankeyData}
      onLinkClick = {
        action('Link has been clicked')
      }
    />
  ))
  .add('without any data', () => <SankeyChart /> )