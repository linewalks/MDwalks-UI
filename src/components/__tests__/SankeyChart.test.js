import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import SankeyChart from '@Charts/SankeyChart';

const SankeyChartData = {
  links: [
    {
      source: 'ER Admission',
      target: 'Re-operation',
      value: 683,
    },
    {
      source: 'Index Invasive Treatment',
      target: 'ER Death',
      value: 44,
    },
    {
      source: 'Re-operation',
      target: 'In Hospital Death',
      value: 84,
    },
    {
      source: 'Re-operation',
      target: 'ER Death',
      value: 9,
    },
  ],
  nodes: [
    {
      name: 'Index Invasive Treatment',
    },
    {
      name: 'In Hospital Death',
    },
    {
      name: 'ER Admission',
    },
    {
      name: 'Re-operation',
    },
    {
      name: 'ER Death',
    },
  ],
}

const changedSankeyChartData = {
  links: [
    {
      source: 'A',
      target: 'B',
      value: 520,
    },
    {
      source: 'B',
      target: 'C',
      value: 37,
    },
    {
      source: 'A',
      target: 'C',
      value: 15,
    },
  ],
  nodes: [
    {
      name: 'A',
    },
    {
      name: 'B',
    },
    {
      name: 'C',
    },
  ],
}

describe('set SelectNodes', () => {
  let wrapper; let selectedNodes;
  beforeEach(() => {
    selectedNodes = []
    wrapper = shallow(<SankeyChart selectedNodes={selectedNodes} data={SankeyChartData} />);
  })

  it('default', () => {
    expect(wrapper.state('selectedNodes')).toEqual([]);
  })

  it('initialize', () => {
    expect(wrapper.state('selectedNodes')).toEqual(selectedNodes);
  })
})

describe('function', () => {
  let wrapper; let instance; let
    onChange
  beforeEach(() => {
    onChange = jest.fn()
    wrapper = mount(<SankeyChart selectedNodes={[]} data={SankeyChartData} onChange={onChange} />)
    instance = wrapper.instance()
  })

  it('selectedNodes', () => {
    wrapper.setState({ selectedNode: ['ER Admission'] })

    instance.setSelectedNode('ER Admission')
    expect(wrapper.state('selectedNodes')).toEqual(['ER Admission'])

    instance.setSelectedNode('Re-operation')
    expect(wrapper.state('selectedNodes')).toEqual(['ER Admission', 'Re-operation'])
  });

  it('createLinkId', () => {
    expect(instance.createLinkId(['a'])).toEqual([])
    expect(instance.createLinkId(['a', 'b'])).toEqual(['aXb'])
    expect(instance.createLinkId(['a', 'b', 'c'])).toEqual(['aXb', 'bXc'])
  })

  it('link connect check false', () => {
    expect(instance.linkConnectCheck('ER Admission', 'Index Invasive Treatment')).toBe(false);
  })

  it('link connect check true', () => {
    expect(instance.linkConnectCheck('ER Admission', 'Re-operation')).toBe(true);
    expect(instance.linkConnectCheck('Re-operation', 'ER Admission')).toBe(true);
    expect(instance.linkConnectCheck('', 'ER Admission')).toBe(true);
  })
})

describe('Selected Snapshot', () => {
  it('default', () => {
    const selectedNodes = []
    const wrapper = mount(<SankeyChart selectedNodes={selectedNodes} data={SankeyChartData} />);
    expect(toJson(wrapper.render())).toMatchSnapshot();
  });

  it('ER Admission to Re-operation', () => {
    const selectedNodes = ['ER Admission', 'Re-operation']
    const wrapper = mount(<SankeyChart selectedNodes={selectedNodes} data={SankeyChartData} />)
    expect(toJson(wrapper.render())).toMatchSnapshot();
  });

  it('Re-operation to ER Admission', () => {
    const selectedNodes = ['Re-operation', 'ER Admission']
    const wrapper = mount(<SankeyChart selectedNodes={selectedNodes} data={SankeyChartData} />);
    expect(toJson(wrapper.render())).toMatchSnapshot();
  });
})

describe('데이터를 가진 Sankey Chart가 데이터 변경시, 변경된 데이터를 반영해야 한다.', () => {
  let wrapper; let instance; let
    onChange

  beforeEach(() => {
    onChange = jest.fn()
    wrapper = mount(<SankeyChart selectedNodes={[]} data={SankeyChartData} onChange={onChange} />)
    instance = wrapper.instance()
  })

  it('변경된 데이터 렌더링', () => {
    const spy = jest.spyOn(instance, 'componentDidUpdate')
    const renderSankeySpy = jest.spyOn(instance, 'renderSankey')
    const removeSankeySpy = jest.spyOn(instance, 'removeSankey')
    wrapper.setProps({
      data: changedSankeyChartData,
    })

    expect(wrapper.props('data')).not.toEqual(SankeyChartData)
    expect(spy).toHaveBeenCalled()
    expect(renderSankeySpy).toHaveBeenCalled()
    expect(removeSankeySpy).toHaveBeenCalled()
  })

  it('데이터를 변경하면 state를 초기화해야된다.', () => {
    const resetSpy = jest.spyOn(instance, 'resetSankey')
    const expectedNode = ['A']
    wrapper.setProps({
      data: changedSankeyChartData,
      selectedNodes: ['A', 'B'],
      defaultdNode: ['A'],
    })

    expect(resetSpy).toHaveBeenCalled()
    expect(wrapper.state('selectedNodes')).toEqual(expectedNode)
  })
})
