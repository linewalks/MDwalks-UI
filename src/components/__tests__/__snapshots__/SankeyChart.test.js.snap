// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selected Snapshot ER Admission to Re-operation 1`] = `
<div>
  <svg
    height="454"
    width="1200"
  >
    <g
      transform="translate(100, 100)"
    >
      <g
        class="nodes"
        font-family="sans-serif"
        font-size="10"
      >
        <g>
          <rect
            height="9.223658872077038"
            id="Index_Invasive_Treatment"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="0"
            y="244.77634112792293"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="12.5"
            y="232.77634112792293"
          >
            Index Invasive Treatment
          </text>
        </g>
        <g>
          <rect
            height="17.60880330123797"
            id="In_Hospital_Death"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="975"
            y="95.43055292598069"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="987.5"
            y="83.43055292598069"
          >
            In Hospital Death
          </text>
        </g>
        <g>
          <rect
            height="143.17634112792297"
            id="ER_Admission"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="0"
            y="0"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="12.5"
            y="-12"
          >
            ER Admission
          </text>
        </g>
        <g>
          <rect
            height="143.1763411279229"
            id="Re-operation"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="487.5"
            y="36.18594496381957"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="500"
            y="24.185944963819573"
          >
            Re-operation
          </text>
        </g>
        <g>
          <rect
            height="11.110316368638252"
            id="ER_Death"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="975"
            y="214.63935622721866"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="987.5"
            y="202.63935622721866"
          >
            ER Death
          </text>
        </g>
      </g>
      <g
        class="links"
        fill="none"
      >
        <g>
          <path
            class="sankey-link"
            d="M25,71.58817056396148C256.25,71.58817056396148,256.25,107.77411552778099,487.5,107.77411552778099"
            id="ER_AdmissionXRe-operation"
            style="stroke-width: 143.1763411279229; opacity: 1; stroke: rgba(24, 155, 255, 0.4);"
          />
        </g>
        <g>
          <path
            class="sankey-link"
            d="M25,249.38817056396144C500,249.38817056396144,500,221.13784315981837,975,221.13784315981837"
            id="Index_Invasive_TreatmentXER_Death"
            style="stroke-width: 9.223658872077024; opacity: 0.04; stroke: #000000;"
          />
        </g>
        <g>
          <path
            class="sankey-link"
            d="M512.5,44.99034661443853C743.75,44.99034661443853,743.75,104.23495457659965,975,104.23495457659965"
            id="Re-operationXIn_Hospital_Death"
            style="stroke-width: 17.608803301237955; opacity: 0.04; stroke: #000000;"
          />
        </g>
        <g>
          <path
            class="sankey-link"
            d="M512.5,54.73807701333813C743.75,54.73807701333813,743.75,215.58268497549923,975,215.58268497549923"
            id="Re-operationXER_Death"
            style="stroke-width: 1.8866574965612095; opacity: 0.04; stroke: #000000;"
          />
        </g>
      </g>
    </g>
  </svg>
</div>
`;

exports[`Selected Snapshot Re-operation to ER Admission 1`] = `
<div>
  <svg
    height="454"
    width="1200"
  >
    <g
      transform="translate(100, 100)"
    >
      <g
        class="nodes"
        font-family="sans-serif"
        font-size="10"
      >
        <g>
          <rect
            height="9.223658872077038"
            id="Index_Invasive_Treatment"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="0"
            y="244.77634112792293"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="12.5"
            y="232.77634112792293"
          >
            Index Invasive Treatment
          </text>
        </g>
        <g>
          <rect
            height="17.60880330123797"
            id="In_Hospital_Death"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="975"
            y="95.43055292598069"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="987.5"
            y="83.43055292598069"
          >
            In Hospital Death
          </text>
        </g>
        <g>
          <rect
            height="143.17634112792297"
            id="ER_Admission"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="0"
            y="0"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="12.5"
            y="-12"
          >
            ER Admission
          </text>
        </g>
        <g>
          <rect
            height="143.1763411279229"
            id="Re-operation"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="487.5"
            y="36.18594496381957"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="500"
            y="24.185944963819573"
          >
            Re-operation
          </text>
        </g>
        <g>
          <rect
            height="11.110316368638252"
            id="ER_Death"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="975"
            y="214.63935622721866"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="987.5"
            y="202.63935622721866"
          >
            ER Death
          </text>
        </g>
      </g>
      <g
        class="links"
        fill="none"
      >
        <g>
          <path
            class="sankey-link"
            d="M25,71.58817056396148C256.25,71.58817056396148,256.25,107.77411552778099,487.5,107.77411552778099"
            id="ER_AdmissionXRe-operation"
            style="stroke-width: 143.1763411279229; opacity: 1; stroke: rgba(255, 58, 31, 0.4);"
          />
        </g>
        <g>
          <path
            class="sankey-link"
            d="M25,249.38817056396144C500,249.38817056396144,500,221.13784315981837,975,221.13784315981837"
            id="Index_Invasive_TreatmentXER_Death"
            style="stroke-width: 9.223658872077024; opacity: 0.04; stroke: #000000;"
          />
        </g>
        <g>
          <path
            class="sankey-link"
            d="M512.5,44.99034661443853C743.75,44.99034661443853,743.75,104.23495457659965,975,104.23495457659965"
            id="Re-operationXIn_Hospital_Death"
            style="stroke-width: 17.608803301237955; opacity: 0.04; stroke: #000000;"
          />
        </g>
        <g>
          <path
            class="sankey-link"
            d="M512.5,54.73807701333813C743.75,54.73807701333813,743.75,215.58268497549923,975,215.58268497549923"
            id="Re-operationXER_Death"
            style="stroke-width: 1.8866574965612095; opacity: 0.04; stroke: #000000;"
          />
        </g>
      </g>
    </g>
  </svg>
</div>
`;

exports[`Selected Snapshot default 1`] = `
<div>
  <svg
    height="454"
    width="1200"
  >
    <g
      transform="translate(100, 100)"
    >
      <g
        class="nodes"
        font-family="sans-serif"
        font-size="10"
      >
        <g>
          <rect
            height="9.223658872077038"
            id="Index_Invasive_Treatment"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="0"
            y="244.77634112792293"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="12.5"
            y="232.77634112792293"
          >
            Index Invasive Treatment
          </text>
        </g>
        <g>
          <rect
            height="17.60880330123797"
            id="In_Hospital_Death"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="975"
            y="95.43055292598069"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="987.5"
            y="83.43055292598069"
          >
            In Hospital Death
          </text>
        </g>
        <g>
          <rect
            height="143.17634112792297"
            id="ER_Admission"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="0"
            y="0"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="12.5"
            y="-12"
          >
            ER Admission
          </text>
        </g>
        <g>
          <rect
            height="143.1763411279229"
            id="Re-operation"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="487.5"
            y="36.18594496381957"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="500"
            y="24.185944963819573"
          >
            Re-operation
          </text>
        </g>
        <g>
          <rect
            height="11.110316368638252"
            id="ER_Death"
            rx="4"
            ry="4"
            style="fill: #002d4f; cursor: pointer;"
            width="25"
            x="975"
            y="214.63935622721866"
          />
          <text
            dy="0.35em"
            fill="rgba(0, 0, 0, 0.4)"
            font-famliy="Spoqa Han Sans"
            font-size="14"
            font-stretch="normal"
            font-style="normal"
            font-weight="normal"
            letter-spacing="0.5px"
            line-height="normal"
            text-anchor="middle"
            x="987.5"
            y="202.63935622721866"
          >
            ER Death
          </text>
        </g>
      </g>
      <g
        class="links"
        fill="none"
      >
        <g>
          <path
            class="sankey-link"
            d="M25,71.58817056396148C256.25,71.58817056396148,256.25,107.77411552778099,487.5,107.77411552778099"
            id="ER_AdmissionXRe-operation"
            style="stroke-width: 143.1763411279229; opacity: 0.04; stroke: #000000;"
          />
        </g>
        <g>
          <path
            class="sankey-link"
            d="M25,249.38817056396144C500,249.38817056396144,500,221.13784315981837,975,221.13784315981837"
            id="Index_Invasive_TreatmentXER_Death"
            style="stroke-width: 9.223658872077024; opacity: 0.04; stroke: #000000;"
          />
        </g>
        <g>
          <path
            class="sankey-link"
            d="M512.5,44.99034661443853C743.75,44.99034661443853,743.75,104.23495457659965,975,104.23495457659965"
            id="Re-operationXIn_Hospital_Death"
            style="stroke-width: 17.608803301237955; opacity: 0.04; stroke: #000000;"
          />
        </g>
        <g>
          <path
            class="sankey-link"
            d="M512.5,54.73807701333813C743.75,54.73807701333813,743.75,215.58268497549923,975,215.58268497549923"
            id="Re-operationXER_Death"
            style="stroke-width: 1.8866574965612095; opacity: 0.04; stroke: #000000;"
          />
        </g>
      </g>
    </g>
  </svg>
</div>
`;
